### PR TITLE
upload avatar in account creation + extract modals to shared js file

### PR DESCRIPTION
**Pull request by brucexhan**

### CODE REVIEW

File: apps/home/home.html [hunk 1]
```diff
@@ -145,77 +145,10 @@
...
-	        </button>
-	      </div>
-	      <div class="modal-body">
-	        <p>Please check your Internet connection. Or blockchain is unavailable.</p>
-	      </div>
-	      <div class="modal-footer">
-	        <button type="button" class="btn btn-secondary" data-dismiss="modal" data-close-modal-id="connectionModal" id="connectionBRClose">Close</button>
-	      </div>
-	    </div>
-	  </div>
```
Comment (brucexhan) [inline apps/home/home.html]: move these three popup modals to shared home.js, then all the home app html pages can get rid of these duplicate HTML.

File: apps/home/home.html [hunk 2]
```diff
@@ -234,37 +167,45 @@ <h5 class="modal-title">Warning</h5>
...
+											setupUserDiv(accountName);
+											setupYourApps();											
+										}
+								    });
+								}
+								else {
+									displayConnectionModal();
+								}
+							}
+					    });
```
Comment (brucexhan) [inline apps/home/home.html]: enable logic after loading wolkjs: 
first is to check if the page can connect to the node by calling getLatestBlockNumber. If not, the modal saying connection err shows up.
second is to check if the account stored in indexeddb exists on the connected node/chain. If not, the modal saying "account doesn't exist" shows up.
If it passes the 2 checks, do the setup process.

File: apps/home/home.js [hunk 3]
```diff
@@ -41,121 +40,280 @@ var navbar = link[0].import.querySelector('.navbar');
...
+	        <p>Please check your Internet connection. Or blockchain is unavailable.</p>
+	      </div>
+	      <div class="modal-footer">
+	        <button type="button" class="btn btn-secondary" data-dismiss="modal" data-close-modal-id="connectionModal" id="connectionBRClose">Close</button>
+	      </div>
+	    </div>
+	  </div>
+	</div>
+
+`;
```
Comment (brucexhan) [inline apps/home/home.js]: declare the three html modals mentioned above as string

File: apps/home/home.js [hunk 4]
```diff
@@ -41,121 +40,280 @@ var navbar = link[0].import.querySelector('.navbar');
...
+======= helper functions ==========
+*/
+
+function insertAfter(el, referenceNode) {
+    referenceNode.parentNode.insertBefore(el, referenceNode.nextSibling);
+}
+
+function htmlStringToElement(str) {
+	return document.createRange().createContextualFragment(str);
+}
```
Comment (brucexhan) [inline apps/home/home.js]: two helper functions:
insertAfter is to append an html element after one other.
htmlStringToElement is to convert html string to html elements. 

these two's main purpose is to convert the three modal in string to html elements then append them after the footer div. Then they will function the same as they are in the html file.

File: apps/home/home.js [hunk 5]
```diff
@@ -41,121 +40,280 @@ var navbar = link[0].import.querySelector('.navbar');
...
 function closeModalByIdReloadEventHandler(e) {
-  var closeModalId = e.target.dataset.closeModalId;
+    var closeModalId = e.target.dataset.closeModalId;
     document.getElementById(closeModalId).style.display = "none";
     location.reload();
 }
 
 function openAddUserModalEventHandler() {
-  clearInputErr("createAccountNameInput");
+	insertAfter(htmlStringToElement(createAccountModal), document.getElementById("footer"));
```
Comment (brucexhan) [inline apps/home/home.js]: as mentioned earlier, this converts createAccountModal string to html element first, then append it after footer.

File: apps/home/home.js [hunk 6]
```diff
@@ -41,121 +40,280 @@ var navbar = link[0].import.querySelector('.navbar');
...
-
-
-      document.getElementById("createAccountResultDiv").innerHTML = `<p><i class='fa fa-spinner fa-spin spinner'></i> Creating... <a href=${link} target="_blank">View on Block Explorer</a></p>`;
-
-      wolkjs.waitForTx(txhash, function(err1, tx) {
-            if (err1) {
-              console.log(err1);
-            }
-            console.log("===> create account TX: ", txhash, tx);
+	insertAfter(htmlStringToElement(connectionModal), document.getElementById("footer"));
```
Comment (brucexhan) [inline apps/home/home.js]: similar as above

File: apps/home/home.js [hunk 7]
```diff
@@ -41,121 +40,280 @@ var navbar = link[0].import.querySelector('.navbar');
...
 
-          document.getElementById("createAccountResultDiv").innerHTML = `<p>Done. <a href="${link}" target='_blank'>View on Block Explorer</a></p>`;
+    document.getElementById("connectionModal").style.display = "block";
+    document.getElementById("connectionTRClose").addEventListener('click', closeModalByIdLightEventHandler, false);
+    document.getElementById("connectionBRClose").addEventListener('click', closeModalByIdLightEventHandler, false);
+}
+
+function displayDeleteAccountModal() {
+	insertAfter(htmlStringToElement(deleteAccountModal), document.getElementById("footer"));
```
Comment (brucexhan) [inline apps/home/home.js]: similar as above

File: apps/home/home.js [hunk 8]
```diff
@@ -41,121 +40,280 @@ var navbar = link[0].import.querySelector('.navbar');
...
 
+    if (accountName == "") {
+        displayInputErr("createAccountNameInput", "createAccountResultDiv");
+        return;
     }
-  });
+    if (!profileImage) {
+		displayInputErr("", "createAccountResultDiv", "No Profile Image Selected!");
+		return;
+	}
```
Comment (brucexhan) [inline apps/home/home.js]: validate account name input and profile image file input before sending the txn, similar to uploading new video/publishing new app in last PR.

File: apps/home/home.js [hunk 9]
```diff
@@ -41,121 +40,280 @@ var navbar = link[0].import.querySelector('.navbar');
...
+		    var profileImageMeta = JSON.stringify({
+		        "@context": "http://schema.org",
+		        "@type": "ImageObject",
+		        "identifier": `${accountName}/${ImageObjectBucket}/${profileImageName}`,
+		        "author": `wolk://${accountName}`,
+		        "name": profileImageName,
+		        "uploadDate": "2019-07-22 09:00:00"
+		    });
+
+	      	wolkjs.waitForTx(txhash, function(err0, tx0) {
```
Comment (brucexhan) [inline apps/home/home.js]: after createAccount txn  is finalized then call setKey for the profile image

File: apps/home/home.js [hunk 10]
```diff
@@ -41,121 +40,280 @@ var navbar = link[0].import.querySelector('.navbar');
...
+					                document.getElementById("createAccountBRClose").removeEventListener('click', closeModalByIdEventHandler, false);
+					                document.getElementById("createAccountTRClose").addEventListener('click', closeModalByIdReloadEventHandler, false);
+					                document.getElementById("createAccountBRClose").addEventListener('click', closeModalByIdReloadEventHandler, false);
+					                hideActionBtn("createAccountCreateBtn");
+
+							    });							
+							}
+						})	
+					}
+				});
```
Comment (brucexhan) [inline apps/home/home.js]: https://github.com/wolkdb/wolkjs/blob/f7d9aa57c17ecca57cb93a4e2dd975198f30d7fa/apps/video/video.js#L86
event listener seems to not like async/await inside it (cannot make `var txnHashes = await wolkjs.setKeys(txnArray);` work inside createAccountEventHandler function, always getting error), so step back to use nested ways to upload the profile image.

File: apps/home/home.js [hunk 11]
```diff
@@ -41,121 +40,280 @@ var navbar = link[0].import.querySelector('.navbar');
...
 }
 
 /*
 ======= common setup functions ==========
 */
 
 function setupUserDiv(accountName) {
-  document.getElementById("currentAccount").innerHTML = accountName;
+    document.getElementById("currentAccount").innerHTML = accountName;
+    document.getElementById("accountImage").src = `${provider}/${owner}/${ImageObjectBucket}/${owner}_profile.png`;
```
Comment (brucexhan) [inline apps/home/home.js]: attach the profile image from server to the navigation bar

File: apps/home/navbar.html [hunk 12]
```diff
@@ -1,47 +1,45 @@
...
-  </div>
+    <a class="navbar-brand" href="./home.html">Wolk Home</a>
+    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
+        <span class="navbar-toggler-icon"></span>
+    </button>
+    <div class="collapse navbar-collapse" id="navbarSupportedContent">
+        <ul class="navbar-nav mr-auto">
+            <li class="nav-item">
+                <a class="nav-link disabled">
+                    <img src="" width="30" height="30" class="d-inline-block align-top" alt="" id="accountImage">
```
Comment (brucexhan) [inline apps/home/navbar.html]: placeholder for the profile image in the navigation bar

File: apps/home/home.js [hunk 13]
```diff
@@ -41,121 +40,280 @@ var navbar = link[0].import.querySelector('.navbar');
...
+        return;
     }
-  });
+    if (!profileImage) {
+		displayInputErr("", "createAccountResultDiv", "No Profile Image Selected!");
+		return;
+	}
+
+    wolkjs.createAccount(accountName, { WaitForTx: "0" }, function(err, result) {
+        if (err) {
```
Comment (rodneywitcher) [inline apps/home/home.js]: Review and remove { WaitForTx } if you confirm that it is actually not being used.

File: apps/home/home.html [hunk 14]
```diff
@@ -234,37 +167,45 @@ <h5 class="modal-title">Warning</h5>
...
-					 //    });
+
+						wolkjs.getLatestBlockNumber({}, function(err, result) {
+							if (err) {
+								console.log(err);
+								displayConnectionModal()
+							} else {
+								console.log(result);
+								var blockenumber = parseInt(result, 10);
+
```
Comment (rodneywitcher) [inline apps/home/home.html]: let's change to blocknumber instead of blockenumber

**Pull request merged**